sdks/java/harness/src/main/java/org/apache/beam/fn/harness/BeamFnDataReadRunner.java

```diff
@@ -237,8 +232,9 @@ public void trySplit(
     }
 
     synchronized (splittingLock) {
-      // Don't attempt to split if we haven't started.
-      if (!started) {
+      // Don't attempt to split if we are already done since there isn't a meaningful split we can
+      // provide.
+      if (index == stopIndex) {
```

Comment:
should we also check `index != -1`?

Reply:
No, the logic below will allow us to choose a stopIndex even if `registerInputLocation` has not been invoked. See `testSplittingWhenNoElementsProcessed`. I added a variant of that test where we split before `registerInputLocation` happens to get coverage for this case.